File: app/src/main/java/com/appsbysha/ohboy/SayingAdapter.java
package com.appsbysha.ohboy;


import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.text.TextUtils.TruncateAt;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.View.OnLongClickListener;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.appsbysha.ohboy.entities.Saying;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

public class SayingAdapter extends RecyclerView.Adapter<SayingAdapter.SayingHolder> {
    private List<Saying> sayings = new ArrayList<>();
    private String dob;
    private OnItemClickListener listener;
    private OnLongClickListener longClickListener;

    @NonNull
    @Override
    public SayingHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
      View itemView = LayoutInflater.from(parent.getContext())
          .inflate(R.layout.saying_item, parent, false);
      return new SayingHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull SayingHolder holder, int position) {
      Saying currentSaying = sayings.get(position);
      holder.textViewTitle.setText(currentSaying.getTitle());
      if(currentSaying.getTitle().isEmpty())
      {holder.textViewTitle.setVisibility(View.GONE);}
      String sayingDate = currentSaying.getDate();
      holder.textViewDate.setText(sayingDate);
      SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy");
      try {
        Date date = sdf.parse(sayingDate);
        Date childDob = sdf.parse(dob);
        Calendar cal = Calendar.getInstance();
        cal.setTime(date);
        Calendar calDob = Calendar.getInstance();
        calDob.setTime(childDob);
        long msdiff =cal.getTimeInMillis() - calDob.getTimeInMillis();
        cal.setTimeInMillis(msdiff);
        String diffYears = String.valueOf(cal.get(Calendar.YEAR)-1970);
        String diffMonths = String.valueOf(cal.get(Calendar.MONTH));
        String diffDays = String.valueOf(cal.get(Calendar.DAY_OF_MONTH));
        holder.textViewAge.setText(diffYears + " years, " + diffMonths + " months and " + diffDays + " days old");

      } catch (ParseException e) {
        e.printStackTrace();
      }
      if(holder.expand) {
        holder.textViewDescription.setMaxLines(Integer.MAX_VALUE);
        holder.textViewDescription.setEllipsize(null);
      }
      else
      {
        holder.textViewDescription.setMaxLines(3);
        holder.textViewDescription.setEllipsize(TruncateAt.END);
      }
      if(currentSaying.getPhoto() != null) {
        Bitmap bitmap = BitmapFactory
            .decodeByteArray(currentSaying.getPhoto(), 0, currentSaying.getPhoto().length);
        holder.photo.setImageBitmap(bitmap);
      }

    }

    @Override
    public int getItemCount() {
      return sayings.size();
    }

    public void setSayings(List<Saying> sayings, String dob) {
      this.sayings = sayings;
      this.dob = dob;
      notifyDataSetChanged();
    }

    class SayingHolder extends RecyclerView.ViewHolder {

      private TextView textViewTitle;
      private TextView textViewDescription;
      private TextView textViewDate;
      private TextView textViewAge;
      private ImageView photo;
      private boolean expand = false;

      public SayingHolder(final View itemView) {
        super(itemView);
        textViewTitle = itemView.findViewById(R.id.text_view_title);
        textViewDescription = itemView.findViewById(R.id.text_view_description);
        textViewDate = itemView.findViewById(R.id.text_view_date);
        textViewAge = itemView.findViewById(R.id.text_view_dob);
        photo = itemView.findViewById(R.id.child_image);
        itemView.setOnClickListener(new OnClickListener() {
          @Override
          public void onClick(View v) {
           int position = getAdapterPosition();
            if (listener != null && position != RecyclerView.NO_POSITION) {
              listener.onItemClick(sayings.get(position));
            }
            expand = !expand;
          }
        });
        itemView.setOnLongClickListener(new View.OnLongClickListener() {
          @Override
          public boolean onLongClick(View v) {
            int position = getAdapterPosition();
            if (longClickListener != null && position != RecyclerView.NO_POSITION) {
              longClickListener.onLongClick(sayings.get(position));
            }
            return false;
          }
        });

      }
    }

  public Saying getSayingAt(int position)
  {
    return sayings.get(position);
  }

  public interface OnItemClickListener {
    void onItemClick(Saying saying);
  }
  public void setOnItemClickListener(OnItemClickListener listener) {
    this.listener = listener;
  }

  public interface OnLongClickListener{
      void onLongClick(Saying saying);
  }

  public void setLongClickListener(OnLongClickListener longClickListener){
      this.longClickListener = longClickListener;
  }
  }

File: app/src/main/java/com/appsbysha/ohboy/adapters/SayingAdapter.kt
package com.appsbysha.ohboy.adapters

import android.app.Activity
import android.content.Context
import android.os.Build
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.appsbysha.ohboy.R
import com.appsbysha.ohboy.adapters.SayingAdapter.SayingHolder
import com.appsbysha.ohboy.database.OhBoyRepository
import com.appsbysha.ohboy.entities.Line
import com.appsbysha.ohboy.entities.Saying
import com.appsbysha.ohboy.interfaces.DataReadyListener
import java.text.ParseException
import java.text.SimpleDateFormat
import java.time.LocalDate
import java.time.Period
import java.util.*

class SayingAdapter(private val activity: Activity?, private val context: Context) : RecyclerView.Adapter<SayingHolder>() {
    private var sayings: List<Saying> = ArrayList()
    private var dob: String? = null
    private var listener: OnItemClickListener? = null
    private var longClickListener: OnLongClickListener? = null


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SayingHolder {
        val itemView = LayoutInflater.from(parent.context)
                .inflate(R.layout.saying_item, parent, false)
        return SayingHolder(itemView)
    }

    @RequiresApi(Build.VERSION_CODES.O)
    override fun onBindViewHolder(holder: SayingHolder, position: Int) {
        val currentSaying = sayings[position]
        holder.textViewTitle.text = currentSaying.title
        if (currentSaying.title.isNullOrEmpty()) {
            holder.textViewTitle.visibility = View.GONE
        }
        val sayingDate = currentSaying.date
        sayingDate?.let {
            holder.textViewDate.text = sayingDate
            val sdf = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault())
            try {
                val date = sdf.parse(sayingDate)
                val childDob = sdf.parse(dob)
                val cal = Calendar.getInstance()
                cal.time = date
               /* cal.set(Calendar.MILLISECOND, 0);
                cal.set(Calendar.SECOND, 0);
                cal.set(Calendar.MINUTE, 0);
                cal.set(Calendar.HOUR_OF_DAY, 0);*/
                val calDob = Calendar.getInstance()
                calDob.time = childDob
   /*             calDob.set(Calendar.MILLISECOND, 0);
                calDob.set(Calendar.SECOND, 0);
                calDob.set(Calendar.MINUTE, 0);
                calDob.set(Calendar.HOUR_OF_DAY, 0);*/
                val sayingDate: LocalDate = LocalDate.of(cal.get(Calendar.YEAR), cal.get(Calendar.MONTH), cal.get(Calendar.DAY_OF_MONTH))
                val dateOfBirth: LocalDate = LocalDate.of(calDob.get(Calendar.YEAR), calDob.get(Calendar.MONTH), calDob.get(Calendar.DAY_OF_MONTH))
                val msdiff = Period.between(dateOfBirth, sayingDate)
                //cal.timeInMillis = msdiff
                val diffYears = msdiff.years// (cal[Calendar.YEAR] - 1970)
                val diffMonths = msdiff.months//cal[Calendar.MONTH]
                val diffDays = msdiff.days//cal[Calendar.DAY_OF_MONTH]
                val age = (if(diffYears != 0) "$diffYears year" else "") +
                        if(diffYears>1)"s" else "" +
                        (if(diffMonths != 0 && diffDays == 0) " and" else "") +
                        (if(diffMonths != 0) " $diffMonths month" else "") +
                         if(diffMonths>1)"s" else "" +
                        (if(diffDays != 0 && (diffMonths != 0 || diffYears != 0)) " and" else "") +
                         (if(diffDays != 0) " $diffDays day" else "") +
                        if(diffDays>1)"s" else "" +
                        " old"
                holder.textViewAge.text =    age
            } catch (e: ParseException) {
                e.printStackTrace()
            }
        }
        if (holder.expand) { //todo expand rv min/max height
            holder.recyclerView.visibility = View.VISIBLE
            holder.expander.setImageDrawable(context.getDrawable(R.drawable.expand_less_foreground))

        } else {
            holder.recyclerView.visibility = View.GONE
            holder.expander.setImageDrawable(context.getDrawable(R.drawable.expand_more_foreground))


        }
        /*      currentSaying.photo?.let {
                   val bitmap = BitmapFactory
                           .decodeByteArray(currentSaying.photo, 0, it.size)
                   holder.photo.setImageBitmap(bitmap)
               }*/

        //todo add loading bar
        OhBoyRepository(activity?.application, currentSaying.childId, currentSaying.sayingId).getAllLinesView(object : DataReadyListener{
            override fun onSuccess(list: List<Line>) {
                super.onSuccess(list)
                holder.adapter.setLines(list as MutableList<Line>, false)
            }
        })

     holder.recyclerView.adapter = holder.adapter
        holder.recyclerView.layoutManager = LinearLayoutManager(context)
        holder.recyclerView.setHasFixedSize(true)


    }

    override fun getItemCount(): Int {
        return sayings.size
    }

    fun setSayings(sayings: List<Saying>, dob: String?) {
        this.sayings = sayings
        this.dob = dob
        notifyDataSetChanged()
    }

    inner class SayingHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val textViewTitle: TextView = itemView.findViewById(R.id.text_view_title)
        val recyclerView: RecyclerView = itemView.findViewById(R.id.linesRecyclerView)
        val adapter: LineAdapter = LineAdapter()
        val textViewDate: TextView = itemView.findViewById(R.id.text_view_date)
        val textViewAge: TextView = itemView.findViewById(R.id.text_view_dob)
        val expander: ImageView = itemView.findViewById(R.id.expanderIcon)

        //  val photo: ImageView = itemView.findViewById(R.id.child_image)
        var expand = false

        init {
            itemView.setOnClickListener {
                val position = adapterPosition
                listener?.let {
                    if (position != RecyclerView.NO_POSITION) {
                        it.onItemClick(sayings[position])
                    }
                }
                expand = !expand
            }
            itemView.setOnLongClickListener {
                val position = adapterPosition
                longClickListener?.let {
                    if (position != RecyclerView.NO_POSITION) {
                        it.onLongClick(sayings[position])
                    }
                }
                false
            }

        }
    }

    fun getSayingAt(position: Int): Saying {
        return sayings[position]
    }

    interface OnItemClickListener {
        fun onItemClick(saying: Saying?)
    }

    fun setOnItemClickListener(listener: OnItemClickListener?) {
        this.listener = listener
    }

    interface OnLongClickListener {
        fun onLongClick(saying: Saying?)
    }

    fun setLongClickListener(longClickListener: OnLongClickListener?) {
        this.longClickListener = longClickListener
    }
}